### PR TITLE
#3128: Add buildsticker directive

### PR DESCRIPTION
**Pull request by shanzi**
Adding buildsticker directive. Here's a primary result

buildsticker

Is this a desired appearance?

## TODO:
- [x] write tests for this directive

### CODE REVIEW

**Comment by tardyp** [comment]
Looks good. The Li bullet needs to be removed

**Comment by shanzi** [comment]
@tardyp Would you like the buildsticker to `float: left` ?

File: www/base/src/app/common/directives/buildsticker/buildsticker.directive.coffee [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class Buildsticker extends Directive('common')
+    constructor: (RecursionHelper) ->
+        return {
+            replace: true
+            restrict: 'E'
+            scope: {build: '='}
+            templateUrl: 'views/buildsticker.html'
+            compile: RecursionHelper.compile
```
Comment (tardyp) [inline www/base/src/app/common/directives/buildsticker/buildsticker.directive.coffee]: I dont think it needs recursive directive

**Comment by tardyp** [comment]
can't wait for the tests!

**Comment by shanzi** [comment]
@tardyp I meet a problem when writing the test. As an asynchronous is needed for the test, and I found the documentation at [jasmine's document](http://jasmine.github.io/1.3/introduction.html) that there should be two function `runs` and `waitsFor`. But it throws a `ReferenceError: Can't find variable: runs` when testing.

**Comment by shanzi** [comment]
@tardyp Have found the cause of this problem, fixing it!

**Comment by tardyp** [comment]
@shanzi 
ngMock has some special tricks to test asynchronous without actually doing asynchronous.
basically, you use scope.$digest(),    $httpBackend.flush() or $timeout.flush()

It is better to use them. Asynchronous tests in js are very hard to follow.

**Comment by tardyp** [comment]
@shanzi please look at other directive tests

**Comment by tothandras** [comment]
@shanzi take a look at [this](https://github.com/buildbot/buildbot/blob/master/www/base/src/app/common/services/recentStorage/recentstorage.service.spec.coffee).
It uses the done function when the promises are resolved. You can read about it [here](http://jasmine.github.io/2.0/introduction.html#section-Asynchronous_Support).

**Comment by shanzi** [comment]
@tardyp @tothandras Thanks for the instructions. I finished a primary version of test. Please give your comments.

It's too late in China now. I will review your comments after sleep :zzz:

File: www/base/src/app/common/directives/buildsticker/buildsticker.directive.spec.coffee [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        $httpBackend.flush()
+        $httpBackend.expectDataGET('builders/1')
+        element = $compile("<buildsticker build='build'></buildsticker>")($rootScope)
+        $httpBackend.flush()
+        $rootScope.build.started_at = moment.unix() # avoid test failed due to time changes
+        $rootScope.$digest()
+        html = element.html()
+
+        expectHtml = '''
+        <div class="panel-heading no-select">
```
Comment (tardyp) [inline www/base/src/app/common/directives/buildsticker/buildsticker.directive.spec.coffee]: I would prefer to avoid testing raw html like this.

Rather make sure that some imporant characteristics are found (href is created by ui-sref, that kind of thing)

This will avoid to update the test for trivial template changes

**Comment by shanzi** [comment]
@tardyp tests updated, please have a look of it.

You said the li bracket should be removed, would you like the sticker to be `float:left` ?

**Comment by shanzi** [comment]
Looks like this now:

stickers

File: www/base/src/app/home/home.tpl.jade [hunk 3]
```diff
@@ -25,6 +25,5 @@
       .well
         h2 Welcome to buildbot
         h4 {{ builds_running.length }} Build{{ builds_running.length > 1 ? 's' : '' }} running currently
-        ul
-          li(ng-repeat="build in builds_running")
```
Comment (tardyp) [inline www/base/src/app/home/home.tpl.jade]: I think you could rather use this to remove the *
```
ul
  li.unstyled
```

File: www/base/src/app/common/directives/buildsticker/buildsticker.directive.spec.coffee [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        resultSpan = row0.children().eq(0)
+        buildLink = row0.children().eq(1)
+        durationSpan = row1.children().eq(0)
+        startedSpan = row1.children().eq(1)
+        stateSpan = row1.children().eq(2)
+
+        # the link of build should be correct
+        expect(buildLink.attr('href')).toBe('#/builders/2/builds/1')
+
+        # pending state
```
Comment (tardyp) [inline www/base/src/app/common/directives/buildsticker/buildsticker.directive.spec.coffee]: great test!

File: www/base/src/app/home/home.tpl.jade (hunk 3 above)
Comment (shanzi) [inline www/base/src/app/home/home.tpl.jade]: @tardyp so the stickers need not to be `float: right` ?

**Comment by shanzi** [comment]
@tardyp current style:

styles

**Comment by tardyp** [comment]
It looks like there is one pixel missing in the bottom roundings:
image

not sure why..

**Comment by shanzi** [comment]
@tardyp use `panel-body` instead of `panel-heading` on the sticker will fix this. the sticker's background will become white. Is it OK?

**Comment by shanzi** [comment]
The white background looks quite well:

sticker

**Comment by tardyp** [comment]
:+1:

**Comment by sa2ajj** [comment]
Thanks, @shanzi
